aclk/aclk_lws_wss_client.c

```diff
@@ -0,0 +1,340 @@
+	}
+	return new;
+}
+
+static inline void lws_wss_packet_buffer_push(struct lws_wss_packet_buffer **list, struct lws_wss_packet_buffer *item)
+{
+	struct lws_wss_packet_buffer *tail = *list;
+	if(!*list) {
+		*list = item;
+		return;
+	}
+	while(tail->next) {
+		tail = tail->next;
+	}
+	tail->next = item;
+}
+
+static inline struct lws_wss_packet_buffer *lws_wss_packet_buffer_pop(struct lws_wss_packet_buffer **list)
+{
+	struct lws_wss_packet_buffer *ret = *list;
+	if(ret != NULL)
+		*list = ret->next;
+	//ret->next = NULL;
+	return ret;
+}
+
+static inline void lws_wss_packet_buffer_free(struct lws_wss_packet_buffer *item)
+{
+	freez(item->data);
+	freez(item);
+}
+
+static inline void _aclk_lws_wss_read_buffer_clear(struct lws_ring *ringbuffer)
+{
+	size_t elems = lws_ring_get_count_waiting_elements(ringbuffer, NULL);
+	lws_ring_consume(ringbuffer, NULL, NULL, elems);
+}
+
+static inline void _aclk_lws_wss_write_buffer_clear(struct lws_wss_packet_buffer **list)
+{
+	struct lws_wss_packet_buffer *i;
+	while((i = lws_wss_packet_buffer_pop(list)) != NULL) {
+		lws_wss_packet_buffer_free(i);
+	}
+	*list = NULL;
+}
+
+static inline void aclk_lws_wss_clear_io_buffers(struct aclk_lws_wss_engine_instance *inst)
+{
+	aclk_lws_mutex_lock(&inst->read_buf_mutex);
+	_aclk_lws_wss_read_buffer_clear(inst->read_ringbuffer);
+	aclk_lws_mutex_unlock(&inst->read_buf_mutex);
+	aclk_lws_mutex_lock(&inst->write_buf_mutex);
+	_aclk_lws_wss_write_buffer_clear(&inst->write_buffer_head);
+	aclk_lws_mutex_unlock(&inst->write_buf_mutex);
+}
+
+static const struct lws_protocols protocols[] = {
+	{
+		"aclk-wss",
+		aclk_lws_wss_callback,
+		sizeof(struct aclk_lws_wss_perconnect_data),
+		0, 0, 0, 0
+	},
+	{ NULL, NULL, 0, 0, 0, 0, 0 }
+};
+
+static void aclk_lws_wss_log_divert(int level, const char *line) {
+	switch(level){
+		case LLL_ERR:
+			error("Libwebsockets Error: %s", line);
+			break;
+		case LLL_WARN:
+			debug(D_ACLK, "Libwebsockets Warn: %s", line);
+			break;
+		default:
+			error("Libwebsockets try to log with unknown log level (%d), msg: %s", level, line);
+	}
+}
+
+struct aclk_lws_wss_engine_instance* aclk_lws_wss_client_init (const struct aclk_lws_wss_engine_callbacks *callbacks, const char *target_hostname, int target_port) {
+	static int lws_logging_initialized = 0;
+	struct lws_context_creation_info info;
+	struct aclk_lws_wss_engine_instance *inst;
+
+	if(unlikely(!lws_logging_initialized)) {
+		lws_set_log_level(LLL_ERR | LLL_WARN, aclk_lws_wss_log_divert);
+		lws_logging_initialized = 1;
+	}
+
+	if(!callbacks || !target_hostname)
+		return NULL;
+
+	inst = callocz(1, sizeof(struct aclk_lws_wss_engine_instance));
+
+	inst->host = target_hostname;
+	inst->port = target_port;
+
+	memset(&info, 0, sizeof(struct lws_context_creation_info));
+    info.options = LWS_SERVER_OPTION_DO_SSL_GLOBAL_INIT;
+	info.port = CONTEXT_PORT_NO_LISTEN;
+    info.protocols = protocols;
+	info.user = inst;
+	
+	inst->lws_context = lws_create_context(&info);
```

Comment:
Please, add a condition here to verify whether the context was allocated correctly.

Reply:
thank you fixed in PR #7988 where this code belong